### ISSUE TITLE
Identify on Demo's AGS Dakota County Rail layer not returning results

### ISSUE THREAD

**Issue by brentfraser** [geomoose/gm3#392]
Go to GeoMoose Demo site
Turn on AGS Dakota County Rail layer
Click Identify button 
Click on rail feature in map.

No results are shown

**Comment by brentfraser** [comment]
I did some investigation using ERSI's web interface to do a point query::

http://gis2.co.dakota.mn.us/arcgis/rest/services/DCGIS_OL_Transportation/MapServer/16/query?where=&text=&objectIds=&time=&geometry=-10389315.44%2C5584710.51&geometryType=esriGeometryPoint&inSR=102100&spatialRel=esriSpatialRelIntersects&relationParam=&outFields=*&returnGeometry=true&returnTrueCurves=false&maxAllowableOffset=&geometryPrecision=&outSR=102100&returnIdsOnly=false&returnCountOnly=false&orderByFields=&groupByFieldsForStatistics=&outStatistics=&returnZ=false&returnM=false&gdbVersion=&returnDistinctValues=false&resultOffset=&resultRecordCount=&f=html

And that did return a record.  Shaving off the last digit from from the point's x and y coord resulted in no records found.  Yikes!

So we may need to include a radius/buffer distance for the point in our query i/f and submit an envelope geom instead of point geom.  ArcGIS 10 may have more/better options...

**Comment by theduckylittle** [comment]
Is there an extra parameter we should be passing in by default?

**Comment by brentfraser** [comment]
Not that I know of (but I haven't done intense Googling).  I didn't find any solid examples of querying by point.

**Comment by brentfraser** [comment]
I think we may need to construct a small bbox, and use that in the query,  Intersecting with a point could be problematic; intersecting with a small rectangle (size based on zoom/scale) would be more predictable.

**Comment by brentfraser** [comment]
In map.js, around line 488 change 
`query_params.spatialRel = 'esriSpatialRelIntersects';`
to: 
`query_params.spatialRel = 'esriSpatialRelEnvelopeIntersects';`
and the Identify works.

At least it works with query geom= point  and feature geom = line.  Not sure about other combinations.

**Comment by theduckylittle** [comment]
So the real answer is that there is a separate Identify function for AGS, however, the server we are using in the demo _does not_ support that function:

- https://developers.arcgis.com/rest/services-reference/identify-map-service-.htm
- https://gis2.co.dakota.mn.us/arcgis/rest/services/DCGIS_OL_Transportation/MapServer/16/

Using the EnvelopeIntersects only works because you're doing a point-in-bounding-box query which is not likely to give the user the results they're really looking for.

**Comment by brentfraser** [comment]
Yes, that's true.  If the server does not support AGS Identify, a better solution would be to create a  1 or 2 pixel rectangle around the mouse point, convert it to projected coordinates and submit that as selection geometry.  Or just document the problem.

**Comment by brentfraser** [comment]
FYI, GM's Identify does not work for points (as well as lines) in an ags-vector mapsource.  In the short term, perhaps a quick fix would be to allow a "tools" option for a spatial filter (like the select service), around line 45 in identify.js:

 ```JavaScript
   /** Limit the number of selection tools available */
    this.tools = options.tools ? options.tools : {
        'Point': true,
        'default': 'Point'
    };
```
So no change in the Identify work flow for most implementations, but we AGS users could modify our app.js to do:

```JavaScript
    app.registerService('identify', IdentifyService, {
        tools : { 'Box': true, 'Point': true, 'default': 'Box'}
    });
```

**Comment by theduckylittle** [comment]
That's not the right fix. That will break identify on services that _properly_ support identify.

We should use the AGS-version of identify (instead of the generic feature service query). We might also be able to add a pixel-buffer "hack" into AGS Feature Service queries for AGS-servers that aren't correctly configured to support identify.

With regards to the demo, it would be good to find a better example which support the identify server functionality then implement the query code.

**Comment by brentfraser** [comment]
Yes that's true it you mix ags-vector and say mapserver types in the same site.  If the user does a Box select for mapserver layers it would fail with "Error: No valid selection geometry."

**Comment by brentfraser** [comment]
We could try a Dakota County layer; it might work:
https://gis2.co.dakota.mn.us/arcgis/rest/services/DCGIS_OL_Environment/MapServer/identify

**Comment by brentfraser** [comment]
If most AGS Mapservers support /query but not /identify, perhaps the better (more-servers-supported) solution would be to construct a box from the point.

**Comment by brentfraser** [comment]
Even [ESRI's web viewer](https://www.arcgis.com/home/webmap/viewer.html?url=https%3A%2F%2Fgis2.co.dakota.mn.us%2Farcgis%2Frest%2Fservices%2FDCGIS_OL_Environment%2FMapServer&source=sd) doesn't do a /identify.  Looking at Weather Stations, ask for a table then click on a point; it does: 
https://gis2.co.dakota.mn.us/arcgis/rest/services/DCGIS_OL_Environment/MapServer/43/query?f=json&where=(1%3D1)%20AND%20(1%3D1)&returnGeometry=false&spatialRel=esriSpatialRelIntersects&outFields=*&orderByFields=OBJECTID%20ASC&outSR=102100&resultOffset=50&resultRecordCount=50

So:
```JSON
{
    "f": "json",
    "where": "(1=1) AND (1=1)",
    "returnGeometry": "true",
    "spatialRel": "esriSpatialRelIntersects",
    "geometry": "{\"xmin\":-10376396.814536516,\"ymin\":5563023.890259778,\"xmax\":-10375938.192366805,\"ymax\":5563482.512429489,\"spatialReference\":{\"wkid\":102100,\"latestWkid\":3857}}",
    "geometryType": "esriGeometryEnvelope",
    "inSR": "102100",
    "outFields": "*",
    "outSR": "102100"
}
```

**Comment by klassenjs** [comment]
I suggest we support the esri servers that exist in the wild over esri documentation that claims unicorns exist.

In any case this seems like something that should be contained to the ags-vector driver as some kind of flag enabling a query rewrite a point to a small box, using the identify endpoint, or leave it alone, rather than making the identify code complicated to deal with one badly behaving data source type.

**Comment by brentfraser** [comment]
Well that was easier than I thought; in map.js after line 473:
```JavaScript
            if (queryGeometry.type === 'Point'){
                const res = this.map.getView().getResolution() * 2;
                const pt = queryGeometry.coordinates;
                queryGeometry.type = 'Polygon'
                queryGeometry.coordinates = [[
                    [pt[0] + res, pt[1] + res],
                    [pt[0] + res, pt[1] - res],
                    [pt[0] - res, pt[1] - res],
                    [pt[0] - res, pt[1] + res]
                ]]
            }
```

**Comment by brentfraser** [comment]
fixed in v3.6.3-beta
